x/margin/keeper/keeper.go

```diff
@@ -102,3 +105,112 @@ func (k Keeper) DestroyMTP(ctx sdk.Context, symbol string, mtpAddress string) er
 	store.Delete(key)
 	return nil
 }
+
+func (k Keeper) ClpKeeper() types.CLPKeeper {
+	return k.clpKeeper
+}
+
+func (k Keeper) BankKeeper() types.BankKeeper {
+	return k.bankKeeper
+}
+
+func (k Keeper) GetLeverageParam(context sdk.Context) sdk.Uint {
+	panic("implement me")
+}
+
+func (k Keeper) CustodySwap(sentBalance sdk.Uint, sentLiabilities sdk.Uint, receivedBalance sdk.Uint, receivedLiabilities sdk.Uint, sentAmount sdk.Uint) sdk.Uint {
+	/*
+	   calculate swap fee based on math spec
+	   lambda_L = (0,1)
+	   Notice this is NOT a simple hybrid of uniswap model and Thorchain slippaged based model
+	   But a upgraded version that include swap, updating bouding curve (to be inside the old one)
+	   One can think about this as a state jump:
+	*/
+
+	// todo review
+	//receivedAmount := sentAmount.Mul(receivedBalance.Add(receivedLiabilities)).
+	//	Quo(sentAmount.Add(sentBalance).Add(sentLiabilities)).
+	//	Add(sentAmount.Mul(receivedBalance.Add(receivedLiabilities)).
+	//	Mul(sentBalance.Add(sentLiabilities)).Quo(sentAmount.Add(sentBalance).Add(sentLiabilities))).MulUint64(2)
```

Comment:
@AustinoBombino can we get this on the spec as a multi step function using the sdk math functions?

Reply:
@timlind Nice catch, Just got into this and realized that it also probably needs to be more sophisticated to account for our decimal normalization logic.

Reply:
@timlind CustodySwap with sdk math functions and decimal normalization is in this PR: https://github.com/Sifchain/Documentation-Private/pull/2